From fc0d32b56622a5822d985f70d4eb580c7547e04a Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Fri, 13 Feb 2015 13:57:51 +0530
Subject: [PATCH 036/132] usb clock init for octeon to be called from dwc3
 driver.

Commit 59c282e24da655ac9449d6944f6d561eb680fdf4 from
git://git.yoctoproject.org/linux-yocto-3.14

Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Vinita Gupta <vgupta@caviumnetworks.com>
Signed-off-by: Abhishek Paliwal <abhishek.paliwal@aricent.com>
[Original patch taken from OCTEON-SDK 3.1.1-544.]
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 drivers/usb/dwc3/core.c | 39 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index 2bbab3d..f802ee9 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -47,6 +47,14 @@
 
 #include "debug.h"
 
+extern void octeon3_usb_phy_reset(int index);
+extern int xhci_octeon_start(struct platform_device *pdev);
+extern int xhci_octeon_stop(struct platform_device *pdev);
+#ifdef CONFIG_USB_XHCI_HCD_OCTEON
+static u64 xhci_octeon_dma_mask = DMA_BIT_MASK(64);
+#endif
+
+
 /* -------------------------------------------------------------------------- */
 
 void dwc3_set_mode(struct dwc3 *dwc, u32 mode)
@@ -67,6 +75,7 @@ static int dwc3_core_soft_reset(struct dwc3 *dwc)
 {
 	u32		reg;
 	int		ret;
+	int		index;
 
 	/* Before Resetting PHY, put Core in Reset */
 	reg = dwc3_readl(dwc->regs, DWC3_GCTL);
@@ -83,6 +92,10 @@ static int dwc3_core_soft_reset(struct dwc3 *dwc)
 	reg |= DWC3_GUSB2PHYCFG_PHYSOFTRST;
 	dwc3_writel(dwc->regs, DWC3_GUSB2PHYCFG(0), reg);
 
+	index = (((uint64_t)(dwc->regs) & 0x10000000000ull) ? 1:0);
+#ifdef CONFIG_USB_XHCI_HCD_OCTEON
+	octeon3_usb_phy_reset(index);
+#else
 	usb_phy_init(dwc->usb2_phy);
 	usb_phy_init(dwc->usb3_phy);
 	ret = phy_init(dwc->usb2_generic_phy);
@@ -94,6 +107,7 @@ static int dwc3_core_soft_reset(struct dwc3 *dwc)
 		phy_exit(dwc->usb2_generic_phy);
 		return ret;
 	}
+#endif
 	mdelay(100);
 
 	/* Clear USB3 PHY reset */
@@ -571,10 +585,12 @@ err2:
 	dwc3_free_scratch_buffers(dwc);
 
 err1:
+#ifndef CONFIG_USB_XHCI_HCD_OCTEON
 	usb_phy_shutdown(dwc->usb2_phy);
 	usb_phy_shutdown(dwc->usb3_phy);
 	phy_exit(dwc->usb2_generic_phy);
 	phy_exit(dwc->usb3_generic_phy);
+#endif
 
 err0:
 	return ret;
@@ -582,11 +598,13 @@ err0:
 
 static void dwc3_core_exit(struct dwc3 *dwc)
 {
+#ifndef CONFIG_USB_XHCI_HCD_OCTEON
 	dwc3_free_scratch_buffers(dwc);
 	usb_phy_shutdown(dwc->usb2_phy);
 	usb_phy_shutdown(dwc->usb3_phy);
 	phy_exit(dwc->usb2_generic_phy);
 	phy_exit(dwc->usb3_generic_phy);
+#endif
 }
 
 static int dwc3_core_get_phy(struct dwc3 *dwc)
@@ -737,6 +755,12 @@ static int dwc3_probe(struct platform_device *pdev)
 	void __iomem		*regs;
 	void			*mem;
 
+#ifdef CONFIG_USB_XHCI_HCD_OCTEON
+	ret = xhci_octeon_start(pdev);
+	if (ret)
+		return ret;
+#endif
+
 	mem = devm_kzalloc(dev, sizeof(*dwc) + DWC3_ALIGN_MASK, GFP_KERNEL);
 	if (!mem)
 		return -ENOMEM;
@@ -883,7 +907,11 @@ static int dwc3_probe(struct platform_device *pdev)
 	platform_set_drvdata(pdev, dwc);
 
 	if (!dev->dma_mask) {
+#ifdef CONFIG_USB_XHCI_HCD_OCTEON
+		dev->dma_mask = &xhci_octeon_dma_mask;
+#else
 		dev->dma_mask = dev->parent->dma_mask;
+#endif
 		dev->dma_parms = dev->parent->dma_parms;
 		dma_set_coherent_mask(dev, dev->parent->coherent_dma_mask);
 	}
@@ -915,6 +943,7 @@ static int dwc3_probe(struct platform_device *pdev)
 		goto err1;
 	}
 
+#ifndef CONFIG_USB_XHCI_HCD_OCTEON
 	usb_phy_set_suspend(dwc->usb2_phy, 0);
 	usb_phy_set_suspend(dwc->usb3_phy, 0);
 	ret = phy_power_on(dwc->usb2_generic_phy);
@@ -924,6 +953,7 @@ static int dwc3_probe(struct platform_device *pdev)
 	ret = phy_power_on(dwc->usb3_generic_phy);
 	if (ret < 0)
 		goto err3;
+#endif
 
 	ret = dwc3_event_buffers_setup(dwc);
 	if (ret) {
@@ -958,8 +988,10 @@ err3:
 	phy_power_off(dwc->usb2_generic_phy);
 
 err2:
+#ifndef CONFIG_USB_XHCI_HCD_OCTEON
 	usb_phy_set_suspend(dwc->usb2_phy, 1);
 	usb_phy_set_suspend(dwc->usb3_phy, 1);
+#endif
 	dwc3_core_exit(dwc);
 
 err1:
@@ -993,16 +1025,21 @@ static int dwc3_remove(struct platform_device *pdev)
 	dwc3_event_buffers_cleanup(dwc);
 	dwc3_free_event_buffers(dwc);
 
+#ifndef CONFIG_USB_XHCI_HCD_OCTEON
 	usb_phy_set_suspend(dwc->usb2_phy, 1);
 	usb_phy_set_suspend(dwc->usb3_phy, 1);
 	phy_power_off(dwc->usb2_generic_phy);
 	phy_power_off(dwc->usb3_generic_phy);
+#endif
 
 	dwc3_core_exit(dwc);
 
 	pm_runtime_put_sync(&pdev->dev);
 	pm_runtime_disable(&pdev->dev);
 
+#ifdef CONFIG_USB_XHCI_HCD_OCTEON
+	xhci_octeon_stop(pdev);
+#endif
 	return 0;
 }
 
@@ -1028,10 +1065,12 @@ static int dwc3_suspend(struct device *dev)
 	dwc->gctl = dwc3_readl(dwc->regs, DWC3_GCTL);
 	spin_unlock_irqrestore(&dwc->lock, flags);
 
+#ifndef CONFIG_USB_XHCI_HCD_OCTEON
 	usb_phy_shutdown(dwc->usb3_phy);
 	usb_phy_shutdown(dwc->usb2_phy);
 	phy_exit(dwc->usb2_generic_phy);
 	phy_exit(dwc->usb3_generic_phy);
+#endif
 
 	return 0;
 }
-- 
1.9.1

